[GStreamer] Get length from content-length header when available in chunked content encoding

eocanha opened the pull request:
#### dc4698316b36833706bc93aff6ad7a18cb3e4c33
<pre>
[GStreamer] Get length from content-length header when available in chunked content encoding
<a href="https://bugs.webkit.org/show_bug.cgi?id=266691">https://bugs.webkit.org/show_bug.cgi?id=266691</a>

Reviewed by Xabier Rodriguez-Calvar.

Currently, chunked content-encoding media content transfers are considered not to have a
defined length and are treated as live videos (no seek available, the whole content is
downloaded in a monolithic way). However, sometimes the web server specifies a
content-length that, even if not mandatory, gives a hint about the total length. If that
info was used, we could treat the download in a much more flexible way, downloading only the
required pieces on demand, as the player needs them, and also enabling seek.

See: <a href="https://github.com/WebPlatformForEmbedded/WPEWebKit/issues/1257">https://github.com/WebPlatformForEmbedded/WPEWebKit/issues/1257</a>

This patch leverages the information supplied by the content-length header when
content-encoding is set as chunked.

* Source/WebCore/platform/graphics/gstreamer/WebKitWebSourceGStreamer.cpp:
(CachedResourceStreamingClient::responseReceived): Set length with the right information when the circumstances are appropriate.

Canonical link: <a href="https://commits.webkit.org/272355@main">https://commits.webkit.org/272355@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d090869d8be9d4bb5b30a94fcffbb9c86cf50743

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/31403 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/10073 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/33097 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/33896 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/28457 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/32170 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/12419 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/7320 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/28086 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/31739 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/8480 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/28037 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/7302 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/7457 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/27943 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/35238 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/28544 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/28396 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/33596 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/7533 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/5563 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/31436 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/9196 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/7373 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/8226 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/8046 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

philn commented on Source/WebCore/platform/graphics/gstreamer/WebKitWebSourceGStreamer.cpp:
```diff
@@ -1010,6 +1010,15 @@ void CachedResourceStreamingClient::responseReceived(PlatformMediaResource&, con
+
+    // But in some cases, Content-Encoding: chunked can specify the total length. Use it.
+    if (!length && response.httpHeaderFields().contains(HTTPHeaderName::TransferEncoding)
+        && (response.httpHeaderField(HTTPHeaderName::TransferEncoding) == String::fromUTF8("chunked"))) {
```
`"chunked"_s`

calvaris replied:
Check should be case insensitive, afaik

eocanha replied:
Also, it looks like HTTP header values are case insensitive, so I should change it to lowercase before comparing.

philn commented on Source/WebCore/platform/graphics/gstreamer/WebKitWebSourceGStreamer.cpp:
```diff
@@ -1010,6 +1010,15 @@ void CachedResourceStreamingClient::responseReceived(PlatformMediaResource&, con
+    // But in some cases, Content-Encoding: chunked can specify the total length. Use it.
+    if (!length && response.httpHeaderFields().contains(HTTPHeaderName::TransferEncoding)
+        && (response.httpHeaderField(HTTPHeaderName::TransferEncoding) == String::fromUTF8("chunked"))) {
+        auto contentLength = parseInteger<uint64_t>(response.httpHeaderField(HTTPHeaderName::ContentLength));
```
this can be moved to the `if` and no need to call `has_value()` i think?

eocanha replied:
The has_value() is there because the webpage can have the Content-Length header, but a malformed value that may not be parseable. That's why parseInteger() returns an std::optional. In case the optional is empty... I really prefer to ignore whatever default (or even undefined?) value the optional may return in that case.

philn replied:
`has_value` and `bool operator` are equivalent, fwiw. https://en.cppreference.com/w/cpp/utility/optional/operator_bool

eocanha committed:
[GStreamer] Get length from content-length header when available in chunked content encoding
https://bugs.webkit.org/show_bug.cgi?id=266691

Reviewed by Xabier Rodriguez-Calvar.

Currently, chunked content-encoding media content transfers are considered not to have a
defined length and are treated as live videos (no seek available, the whole content is
downloaded in a monolithic way). However, sometimes the web server specifies a
content-length that, even if not mandatory, gives a hint about the total length. If that
info was used, we could treat the download in a much more flexible way, downloading only the
required pieces on demand, as the player needs them, and also enabling seek.

See: https://github.com/WebPlatformForEmbedded/WPEWebKit/issues/1257

This patch leverages the information supplied by the content-length header when
content-encoding is set as chunked.

* Source/WebCore/platform/graphics/gstreamer/WebKitWebSourceGStreamer.cpp:
(CachedResourceStreamingClient::responseReceived): Set length with the right information when the circumstances are appropriate.

Canonical link: https://commits.webkit.org/272355@main
